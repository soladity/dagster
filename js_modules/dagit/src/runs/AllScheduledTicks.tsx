import {gql, useQuery} from '@apollo/client';
import {Colors} from '@blueprintjs/core';
import * as React from 'react';

import {showCustomAlert} from 'src/app/CustomAlertProvider';
import {PYTHON_ERROR_FRAGMENT} from 'src/app/PythonErrorInfo';
import {INSTANCE_HEALTH_FRAGMENT} from 'src/instance/InstanceHealthFragment';
import {SchedulerInfoQuery} from 'src/runs/types/SchedulerInfoQuery';
import {POLL_INTERVAL} from 'src/runs/useCursorPaginatedQuery';
import {REPOSITORY_SCHEDULES_FRAGMENT} from 'src/schedules/ScheduleUtils';
import {SchedulerInfo, SCHEDULER_FRAGMENT} from 'src/schedules/SchedulerInfo';
import {SchedulesNextTicks} from 'src/schedules/SchedulesNextTicks';
import {Alert} from 'src/ui/Alert';
import {ButtonLink} from 'src/ui/ButtonLink';
import {Group} from 'src/ui/Group';
import {Loading} from 'src/ui/Loading';

export const AllScheduledTicks = () => {
  const queryResult = useQuery<SchedulerInfoQuery>(SCHEDULER_INFO_QUERY, {
    fetchPolicy: 'cache-and-network',
    pollInterval: POLL_INTERVAL,
    partialRefetch: true,
  });

  return (
    <Loading queryResult={queryResult}>
      {(result) => {
        const {repositoriesOrError, instance, scheduler} = result;
        if (repositoriesOrError.__typename === 'PythonError') {
          const message = repositoriesOrError.message;
          return (
            <Alert
              intent="warning"
              title={
                <Group direction="row" spacing={4}>
                  <div>Could not load scheduled ticks.</div>
                  <ButtonLink
                    color={Colors.BLUE3}
                    underline="always"
                    onClick={() => {
                      showCustomAlert({
                        title: 'Python error',
                        body: message,
                      });
                    }}
                  >
                    View error
                  </ButtonLink>
                </Group>
              }
            />
          );
        }
        return (
          <Group direction="column" spacing={16}>
            <SchedulerInfo schedulerOrError={scheduler} daemonHealth={instance.daemonHealth} />
            <SchedulesNextTicks repos={repositoriesOrError.nodes} />
          </Group>
        );
      }}
    </Loading>
  );
};

const SCHEDULER_INFO_QUERY = gql`
  query SchedulerInfoQuery {
    scheduler {
      ...SchedulerFragment
    }
    instance {
      ...InstanceHealthFragment
    }
    repositoriesOrError {
      ... on RepositoryConnection {
        nodes {
          __typename
          id
          ... on Repository {
            ...RepositorySchedulesFragment
          }
        }
      }
      ...PythonErrorFragment
    }
  }
  ${SCHEDULER_FRAGMENT}
  ${INSTANCE_HEALTH_FRAGMENT}
  ${REPOSITORY_SCHEDULES_FRAGMENT}
  ${PYTHON_ERROR_FRAGMENT}
`;
