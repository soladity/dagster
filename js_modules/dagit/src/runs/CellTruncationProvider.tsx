import * as React from "react";
import * as ReactDOM from "react-dom";
import styled from "styled-components";
import { Colors } from "@blueprintjs/core";
import { showCustomAlert } from "../CustomAlertProvider";

const OverflowFade = styled.div`
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  height: 40px;
  user-select: none;
  pointer-events: none;
  background: linear-gradient(
    to bottom,
    rgba(245, 248, 250, 0) 0%,
    rgba(245, 248, 250, 255) 100%
  );
`;

const OverflowBanner = styled.div`
  position: absolute;
  bottom: 0;
  right: 105px;
  user-select: none;
  font-size: 12px;
  background: ${Colors.LIGHT_GRAY3};
  border-top-left-radius: 4px;
  padding: 2px 12px;
  color: ${Colors.BLACK};
  &:hover {
    color: ${Colors.BLACK};
    background: ${Colors.LIGHT_GRAY1};
  }
`;

export class CellTruncationProvider extends React.Component<
  { style: React.CSSProperties },
  { isOverflowing: boolean }
> {
  state = {
    isOverflowing: false
  };

  componentDidMount() {
    this.detectOverflow();
  }

  componentDidUpdate() {
    this.detectOverflow();
  }

  detectOverflow() {
    // eslint-disable-next-line react/no-find-dom-node
    const el = ReactDOM.findDOMNode(this);
    if (!el || !(el instanceof HTMLElement)) return;

    const child = el.firstElementChild;
    if (!child) return;

    const isOverflowing = child.scrollHeight > this.props.style.height!;
    if (isOverflowing !== this.state.isOverflowing) {
      this.setState({ isOverflowing });
    }
  }

  onView = () => {
    // eslint-disable-next-line react/no-find-dom-node
    const el = ReactDOM.findDOMNode(this) as HTMLElement;
    const message = el.firstChild && el.firstChild.textContent;
    if (!message) return;
    showCustomAlert({ message: message, pre: true });
  };

  render() {
    const { style } = this.props;

    return (
      <div style={style}>
        {this.props.children}
        {this.state.isOverflowing && (
          <>
            <OverflowFade />
            <OverflowBanner onClick={this.onView}>
              View Full Message
            </OverflowBanner>
          </>
        )}
      </div>
    );
  }
}
