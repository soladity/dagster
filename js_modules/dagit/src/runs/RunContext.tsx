import * as React from 'react';

import {RunFragment} from 'src/runs/types/RunFragment';

const initialContext = undefined;
export const RunContext = React.createContext<RunFragment | undefined>(initialContext);
