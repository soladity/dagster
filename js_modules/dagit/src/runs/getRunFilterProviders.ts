export const getRunFilterProviders = (stepNames: string[] = []) => {
  return [
    {
      token: 'step',
      values: () => stepNames,
    },
    {
      token: 'type',
      values: () => ['expectation', 'materialization', 'engine', 'input', 'output', 'pipeline'],
    },
    {
      token: 'query',
      values: () => [],
    },
  ];
};
