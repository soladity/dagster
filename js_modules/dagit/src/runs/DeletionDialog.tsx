import {useMutation} from '@apollo/client';
import {Button, Classes, Colors, Dialog, Icon, ProgressBar} from '@blueprintjs/core';
import * as React from 'react';

import {NavigationBlock} from 'src/runs/NavitationBlock';
import {DELETE_MUTATION} from 'src/runs/RunUtils';
import {
  Delete,
  Delete_deletePipelineRun_PipelineRunNotFoundError,
  Delete_deletePipelineRun_PythonError,
} from 'src/runs/types/Delete';
import {Group} from 'src/ui/Group';
import {FontFamily} from 'src/ui/styles';

export interface Props {
  isOpen: boolean;
  onClose: () => void;
  onComplete: () => void;
  onTerminateInstead: () => void;
  selectedIDs: string[];
  terminatableIDs: string[];
}

type Error =
  | Delete_deletePipelineRun_PythonError
  | Delete_deletePipelineRun_PipelineRunNotFoundError
  | undefined;

type DeletionDialogState = {
  step: 'initial' | 'deleting' | 'completed';
  deletion: {completed: number; errors: {[id: string]: Error}};
};

type DeletionDialogAction =
  | {type: 'reset'}
  | {type: 'start'}
  | {type: 'deletion-success'}
  | {type: 'deletion-error'; id: string; error: Error}
  | {type: 'complete'};

const initialState: DeletionDialogState = {
  step: 'initial',
  deletion: {completed: 0, errors: {}},
};

type Reducer = (
  prevState: DeletionDialogState,
  action: DeletionDialogAction,
) => DeletionDialogState;

const deletionDialogReducer = (
  prevState: DeletionDialogState,
  action: DeletionDialogAction,
): DeletionDialogState => {
  switch (action.type) {
    case 'reset':
      return initialState;
    case 'start':
      return {...prevState, step: 'deleting'};
    case 'deletion-success': {
      const {deletion} = prevState;
      return {
        ...prevState,
        step: 'deleting',
        deletion: {...deletion, completed: deletion.completed + 1},
      };
    }
    case 'deletion-error': {
      const {deletion} = prevState;
      return {
        ...prevState,
        step: 'deleting',
        deletion: {
          ...deletion,
          completed: deletion.completed + 1,
          errors: {...deletion.errors, [action.id]: action.error},
        },
      };
    }
    case 'complete':
      return {...prevState, step: 'completed'};
  }
};

export const DeletionDialog = (props: Props) => {
  const {isOpen, onClose, onComplete, onTerminateInstead, selectedIDs} = props;
  const [state, dispatch] = React.useReducer<Reducer>(deletionDialogReducer, initialState);

  const count = selectedIDs.length;
  const terminatableCount = props.terminatableIDs.length;

  React.useEffect(() => {
    if (isOpen) {
      dispatch({type: 'reset'});
    }
  }, [isOpen]);

  const [destroy] = useMutation<Delete>(DELETE_MUTATION);

  const mutate = async () => {
    dispatch({type: 'start'});

    for (let ii = 0; ii < selectedIDs.length; ii++) {
      const runId = selectedIDs[ii];
      const {data} = await destroy({variables: {runId}});

      if (data?.deletePipelineRun.__typename === 'DeletePipelineRunSuccess') {
        dispatch({type: 'deletion-success'});
      } else {
        dispatch({type: 'deletion-error', id: runId, error: data?.deletePipelineRun});
      }
    }

    dispatch({type: 'complete'});
    onComplete();
  };

  const progressContent = () => {
    switch (state.step) {
      case 'initial':
        return (
          <Group direction="vertical" spacing={8}>
            <div>{`${count} ${count === 1 ? 'run' : 'runs'} will be deleted.`}</div>
            <div>
              Deleting runs will not prevent them from continuing to execute, and may result in
              unexpected behavior.
            </div>
            {terminatableCount ? (
              <div>
                {terminatableCount > 1 ? (
                  <>
                    {`${terminatableCount} of these runs can be terminated. `}
                    <strong>
                      Please consider terminating these runs instead of deleting them.
                    </strong>
                  </>
                ) : (
                  <>
                    {`1 run can be terminated instead. `}
                    <strong>Please consider terminating this run instead of deleting it.</strong>
                  </>
                )}
              </div>
            ) : null}
            <div>Do you wish to continue with deletion?</div>
          </Group>
        );
      case 'deleting':
      case 'completed':
        const value = count > 0 ? state.deletion.completed / count : 1;
        return (
          <Group direction="vertical" spacing={8}>
            <div>Deleting…</div>
            <ProgressBar intent="primary" value={Math.max(0.1, value)} animate={value < 1} />
            {state.step === 'deleting' ? (
              <NavigationBlock message="Deletion in progress, please do not navigate away yet." />
            ) : null}
          </Group>
        );
      default:
        return null;
    }
  };

  const buttons = () => {
    switch (state.step) {
      case 'initial':
        return (
          <>
            <Button intent="none" onClick={onClose}>
              Cancel
            </Button>
            <Button intent="danger" onClick={mutate}>
              {`Yes, delete ${`${count} ${count === 1 ? 'run' : 'runs'}`}`}
            </Button>
            {terminatableCount ? (
              <Button intent="primary" onClick={onTerminateInstead}>
                {`Terminate ${`${terminatableCount} ${
                  terminatableCount === 1 ? 'run' : 'runs'
                }`} instead`}
              </Button>
            ) : null}
          </>
        );
      case 'deleting':
        return (
          <Button intent="danger" disabled>
            Deleting…
          </Button>
        );
      case 'completed':
        return (
          <Button intent="primary" onClick={onClose}>
            Done
          </Button>
        );
    }
  };

  const completionContent = () => {
    if (state.step === 'initial') {
      return null;
    }

    if (state.step === 'deleting') {
      return <div>Please do not close the window or navigate away during deletion.</div>;
    }

    const errors = state.deletion.errors;
    const errorCount = Object.keys(errors).length;
    const successCount = state.deletion.completed - errorCount;

    return (
      <Group direction="vertical" spacing={8}>
        {successCount ? (
          <Group direction="horizontal" spacing={8} alignItems="flex-start">
            <Icon icon="tick-circle" iconSize={16} color={Colors.GREEN3} />
            <div>{`Successfully deleted ${successCount} ${
              successCount === 1 ? 'run' : 'runs'
            }.`}</div>
          </Group>
        ) : null}
        {errorCount ? (
          <Group direction="vertical" spacing={8}>
            <Group direction="horizontal" spacing={8} alignItems="flex-start">
              <Icon icon="warning-sign" iconSize={16} color={Colors.GOLD3} />
              <div>{`Could not delete ${errorCount} ${errorCount === 1 ? 'run' : 'runs'}.`}</div>
            </Group>
            <ul>
              {Object.keys(errors).map((runId) => (
                <li key={runId}>
                  <Group direction="horizontal" spacing={8}>
                    <span style={{fontFamily: FontFamily.monospace}}>{runId.slice(0, 8)}</span>
                    {errors[runId] ? <div>{errors[runId]?.message}</div> : null}
                  </Group>
                </li>
              ))}
            </ul>
          </Group>
        ) : null}
      </Group>
    );
  };

  const canQuicklyClose = state.step !== 'deleting';

  return (
    <Dialog
      isOpen={isOpen}
      title="Delete runs"
      canEscapeKeyClose={canQuicklyClose}
      canOutsideClickClose={canQuicklyClose}
      isCloseButtonShown={canQuicklyClose}
      onClose={onClose}
    >
      <div className={Classes.DIALOG_BODY}>
        <Group direction="vertical" spacing={24}>
          {progressContent()}
          {completionContent()}
        </Group>
      </div>
      <div className={Classes.DIALOG_FOOTER}>
        <div className={Classes.DIALOG_FOOTER_ACTIONS}>{buttons()}</div>
      </div>
    </Dialog>
  );
};
