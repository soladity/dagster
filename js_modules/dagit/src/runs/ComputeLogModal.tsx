import {gql} from '@apollo/client';
import {Dialog} from '@blueprintjs/core';
import * as React from 'react';
import styled from 'styled-components/macro';

import {DirectGraphQLSubscription} from 'src/app/DirectGraphQLSubscription';
import {ComputeLogContent, COMPUTE_LOG_CONTENT_FRAGMENT} from 'src/runs/ComputeLogContent';
import {RunContext} from 'src/runs/RunContext';
import {IStepState} from 'src/runs/RunMetadataProvider';
import {ComputeLogContentFileFragment} from 'src/runs/types/ComputeLogContentFileFragment';
import {ComputeLogsSubscription} from 'src/runs/types/ComputeLogsSubscription';
import {ComputeLogsSubscriptionFragment} from 'src/runs/types/ComputeLogsSubscriptionFragment';
import {ComputeIOType} from 'src/types/globalTypes';
import {Spinner} from 'src/ui/Spinner';

const MAX_STREAMING_LOG_BYTES = 5242880; // 5 MB

interface IComputeLogLink {
  children: React.ReactNode;
  runState: IStepState;
  stepKey: string;
}

export const ComputeLogLink = ({runState, stepKey, children}: IComputeLogLink) => {
  const [isOpen, setOpen] = React.useState(false);
  const run = React.useContext(RunContext);

  if (!run || !run.runId || runState === IStepState.SKIPPED) {
    return null;
  }

  const open = (e: React.SyntheticEvent) => {
    e.stopPropagation();
    setOpen(true);
  };
  const close = () => setOpen(false);
  return (
    <>
      <span onClick={open}>{children}</span>
      <Dialog
        onClose={close}
        style={{
          width: '100vw',
          height: '100vh',
          margin: 0,
          padding: 0,
          borderRadius: 0,
        }}
        usePortal={true}
        isOpen={isOpen}
      >
        {isOpen ? (
          <ComputeLogModal
            runId={run.runId}
            runState={runState}
            stepKey={stepKey}
            onRequestClose={close}
          />
        ) : (
          <LoadingContainer>
            <Spinner purpose="section" />
          </LoadingContainer>
        )}
      </Dialog>
    </>
  );
};

interface ComputeLogModalProps {
  runId: string;
  stepKey: string;
  runState: IStepState;
  onRequestClose: () => void;
}

const ComputeLogModal = ({runId, onRequestClose, stepKey, runState}: ComputeLogModalProps) => {
  return (
    <ComputeLogsProvider runId={runId} stepKey={stepKey} maxBytes={MAX_STREAMING_LOG_BYTES}>
      {({isLoading, stdout, stderr, maxBytes}) => {
        if (isLoading) {
          return (
            <LoadingContainer>
              <Spinner purpose="section" />
            </LoadingContainer>
          );
        }

        return (
          <ComputeLogContent
            runState={runState}
            onRequestClose={onRequestClose}
            stdout={stdout}
            stderr={stderr}
            maxBytes={maxBytes}
          />
        );
      }}
    </ComputeLogsProvider>
  );
};

interface IComputeLogsProviderProps {
  children: (props: {
    isLoading: boolean;
    stdout: ComputeLogsSubscriptionFragment | null;
    stderr: ComputeLogsSubscriptionFragment | null;
    maxBytes: number;
  }) => React.ReactChild;
  runId: string;
  stepKey: string;
  maxBytes: number;
}
interface IComputeLogsProviderState {
  stdout: ComputeLogsSubscriptionFragment | null;
  stderr: ComputeLogsSubscriptionFragment | null;
  isLoading: boolean;
}

class ComputeLogsProvider extends React.Component<
  IComputeLogsProviderProps,
  IComputeLogsProviderState
> {
  _stdout: DirectGraphQLSubscription<ComputeLogsSubscription>;
  _stderr: DirectGraphQLSubscription<ComputeLogsSubscription>;
  state: IComputeLogsProviderState = {
    stdout: null,
    stderr: null,
    isLoading: true,
  };

  componentDidMount() {
    this.subscribe();
  }

  componentWillUnmount() {
    this.unsubscribe();
  }

  componentDidUpdate(prevProps: IComputeLogsProviderProps) {
    if (prevProps.runId !== this.props.runId || prevProps.stepKey !== this.props.stepKey) {
      this.unsubscribe();
      this.subscribe();
    }
  }

  subscribe() {
    const {runId, stepKey} = this.props;
    this.setState({isLoading: true});
    this._stdout = new DirectGraphQLSubscription<ComputeLogsSubscription>(
      COMPUTE_LOGS_SUBSCRIPTION,
      {runId, stepKey, ioType: ComputeIOType.STDOUT, cursor: null},
      this.onStdout,
      this.onError,
    );
    this._stderr = new DirectGraphQLSubscription<ComputeLogsSubscription>(
      COMPUTE_LOGS_SUBSCRIPTION,
      {runId, stepKey, ioType: ComputeIOType.STDERR, cursor: null},
      this.onStderr,
      this.onError,
    );
  }

  unsubscribe() {
    if (this._stdout) {
      this._stdout.close();
    }
    if (this._stderr) {
      this._stderr.close();
    }
  }

  onStdout = (messages: ComputeLogsSubscription[], _: boolean) => {
    this.onMessages('stdout', messages);
  };

  onStderr = (messages: ComputeLogsSubscription[], _: boolean) => {
    this.onMessages('stderr', messages);
  };

  onMessages = (ioType: string, messages: ComputeLogsSubscription[]) => {
    let computeLogs = this.state[ioType];
    messages.forEach((subscription: ComputeLogsSubscription) => {
      computeLogs = this.merge(computeLogs, subscription.computeLogs);
    });

    if (ioType === 'stdout') {
      this.setState({stdout: computeLogs, isLoading: false});
    } else {
      this.setState({stderr: computeLogs, isLoading: false});
    }
  };

  onError = () => {
    this.setState({isLoading: false});
  };

  merge(a: ComputeLogContentFileFragment | null, b: ComputeLogContentFileFragment | null) {
    if (!b) {
      return a;
    }
    let data = a?.data;
    if (a?.data && b?.data) {
      data = this.slice(a.data + b.data);
    } else if (b?.data) {
      data = this.slice(b.data);
    }
    return {
      __typename: b.__typename,
      path: b.path,
      downloadUrl: b.downloadUrl,
      data: data,
      cursor: b.cursor,
    };
  }

  slice(s: string) {
    if (s.length < MAX_STREAMING_LOG_BYTES) {
      return s;
    }
    return s.slice(-MAX_STREAMING_LOG_BYTES);
  }

  render() {
    const {isLoading, stdout, stderr} = this.state;
    const {maxBytes} = this.props;
    return this.props.children({isLoading, stdout, stderr, maxBytes});
  }
}

export const COMPUTE_LOGS_SUBSCRIPTION_FRAGMENT = gql`
  fragment ComputeLogsSubscriptionFragment on ComputeLogFile {
    data
    cursor
    ...ComputeLogContentFileFragment
  }
  ${COMPUTE_LOG_CONTENT_FRAGMENT}
`;

const COMPUTE_LOGS_SUBSCRIPTION = gql`
  subscription ComputeLogsSubscription(
    $runId: ID!
    $stepKey: String!
    $ioType: ComputeIOType!
    $cursor: String
  ) {
    computeLogs(runId: $runId, stepKey: $stepKey, ioType: $ioType, cursor: $cursor) {
      ...ComputeLogsSubscriptionFragment
    }
  }
  ${COMPUTE_LOGS_SUBSCRIPTION_FRAGMENT}
`;

const LoadingContainer = styled.div`
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
`;
