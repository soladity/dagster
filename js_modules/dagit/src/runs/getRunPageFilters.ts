import {tokenizedValueFromString} from 'src/TokenizingField';
import {DefaultLogLevels, LogLevel} from 'src/runs/LogLevel';
import {LogFilter} from 'src/runs/LogsProvider';
import {getRunFilterProviders} from 'src/runs/getRunFilterProviders';

/**
 * Query parameter structure, all optional:
 *
 * `steps`
 *   - string (selection syntax)
 *   - Initializes step selection in Gaant chart
 * `logs`
 *   - string (comma-separated foo:bar tokens or free text)
 *   - Initializes log filter input
 * `levels`
 *   - string (comma-separated values)
 *   - Initializes levels in log filter
 */
export const getRunPageFilters = (searchValue: string) => {
  const params = new URLSearchParams(searchValue);

  const stepQuery = params.get('steps') || '*';
  const logValues = params.getAll('logs') || [];
  const focusedTime = params.get('timestamp') ? Number(params.get('timestamp')) : null;
  const providers = getRunFilterProviders();
  const logQuery = logValues.map((token) => tokenizedValueFromString(token, providers));

  const levelsValues = params.getAll('levels');
  const levelsArr = levelsValues.length ? levelsValues : DefaultLogLevels;
  const levels = levelsArr
    .map((level) => level.toUpperCase())
    .filter((level) => LogLevel.hasOwnProperty(level))
    .reduce((accum, level) => ({...accum, [level]: true}), {});

  return {
    sinceTime: 0,
    focusedTime,
    stepQuery,
    logQuery,
    levels,
  } as LogFilter;
};
