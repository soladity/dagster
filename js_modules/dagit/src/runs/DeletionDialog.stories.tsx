import {Story, Meta} from '@storybook/react/types-6-0';
import faker from 'faker';
import * as React from 'react';

import {ApolloTestProvider} from '../testing/ApolloTestProvider';

import {DeletionDialog, Props as DeletionDialogProps} from './DeletionDialog';

// eslint-disable-next-line import/no-default-export
export default {
  title: 'DeletionDialog',
  component: DeletionDialog,
} as Meta;

const Template: Story<DeletionDialogProps & {mocks?: any}> = ({mocks, ...props}) => (
  <ApolloTestProvider mocks={mocks}>
    <DeletionDialog {...props} />
  </ApolloTestProvider>
);

const ids = [
  faker.random.uuid().slice(0, 8),
  faker.random.uuid().slice(0, 8),
  faker.random.uuid().slice(0, 8),
];

export const Success = Template.bind({});
Success.args = {
  isOpen: true,
  onClose: () => {
    console.log('Close!');
  },
  onTerminateInstead: () => {
    console.log('Terminate instead!');
  },
  selectedRuns: ids.reduce((accum, id) => ({...accum, [id]: true}), {}),
  mocks: {
    Mutation: () => ({
      deletePipelineRun: () => ({
        __typename: 'DeletePipelineRunSuccess',
      }),
    }),
  },
};

export const WithError = Template.bind({});
WithError.args = {
  isOpen: true,
  onClose: () => {
    console.log('Close!');
  },
  onTerminateInstead: () => {
    console.log('Terminate instead!');
  },
  selectedRuns: ids.reduce((accum, id) => ({...accum, [id]: true}), {}),
  mocks: {
    Mutation: () => ({
      deletePipelineRun: (args: {runId: string}) => {
        // Fail the last run
        if (args.runId === ids[2]) {
          return {
            __typename: 'PythonError',
            message: 'Oh no!',
          };
        }
        return {
          __typename: 'DeletePipelineRunSuccess',
        };
      },
    }),
  },
};
