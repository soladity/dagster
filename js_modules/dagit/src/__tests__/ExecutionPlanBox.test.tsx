import * as React from "react";
import * as TestRenderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import {
  ExecutionPlanBox,
  IExecutionPlanBoxProps,
  IExecutionPlanBoxState
} from "../plan/ExecutionPlanBox";
import {
  IExpectationResult,
  IExpectationResultStatus,
  IMaterialization,
  IStepState,
  IStepDisplayIconType,
  IStepDisplayActionType
} from "../RunMetadataProvider";
import { RunContext } from "../runs/RunContext";
import { MOCKS } from "./AppMocks";
import { MockedProvider } from "./MockedProvider";

it("renders given materializations and passing expectations", () => {
  const tables = [
    "users",
    "groups",
    "events",
    "friends",
    "pages",
    "fans",
    "event_admins",
    "group_admins"
  ];

  const expectationResults: IExpectationResult[] = tables.map(table => ({
    icon: IStepDisplayIconType.SUCCESS,
    items: [],
    status: IExpectationResultStatus.PASSED,
    text: table + ".row_count"
  }));

  const materializations: IMaterialization[] = tables.map(table => ({
    icon: IStepDisplayIconType.LINK,
    text: table,
    items: [
      {
        action: IStepDisplayActionType.COPY,
        actionText: "[Copy Path]",
        actionValue: "/path/to/" + table,
        text: table
      }
    ]
  }));

  const executionPlanBoxProps: IExecutionPlanBoxProps = {
    delay: 200,
    executionArtifactsPersisted: false,
    expectationResults: expectationResults,
    materializations,
    stepKey: "many_materializations_and_passing_expectations",
    state: IStepState.SUCCEEDED,
    start: 1558389791907,
    finish: 1558389791907 + 21
  };

  const executionPlanBoxStateExpanded: IExecutionPlanBoxState = {
    expanded: true,
    v: 0
  };

  const run = undefined;
  const component = TestRenderer.create(
    <BrowserRouter>
      <MockedProvider mocks={MOCKS} addTypename={true}>
        <RunContext.Provider value={run}>
          <ExecutionPlanBox {...executionPlanBoxProps} />
        </RunContext.Provider>
      </MockedProvider>
    </BrowserRouter>
  );
  expect(component.toJSON()).toMatchSnapshot();
  component.root.instance.setState(executionPlanBoxStateExpanded);
  expect(component.toJSON()).toMatchSnapshot();
});
