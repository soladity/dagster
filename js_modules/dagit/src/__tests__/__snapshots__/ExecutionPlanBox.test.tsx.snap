// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders given materializations and passing expectations 1`] = `
<div
  className="ExecutionPlanBox__ExecutionPlanRowContainer-sc-1bacjj9-5 kaLnAx"
>
  <div
    className="ExecutionPlanBox__ExecutionPlanBoxContainer-sc-1bacjj9-6 beJStN succeeded"
    onClick={[Function]}
    style={
      Object {
        "transitionDelay": "200ms",
      }
    }
  >
    <div
      className="ExecutionPlanBox__ExecutionFinishedFlash-sc-1bacjj9-4 bRVdHa"
      style={
        Object {
          "transitionDelay": "200ms",
        }
      }
    />
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "inline-flex",
        }
      }
    >
      <div
        className="ExecutionPlanBox__ExecutionStateWrap-sc-1bacjj9-0 fipiZp"
        onClick={[Function]}
      >
        <div
          className="ExecutionStateDot-bprv14-0 gPIOyE"
          style={
            Object {
              "transitionDelay": "200ms",
            }
          }
          title="Succeeded"
        />
        <div
          style={
            Object {
              "width": 4,
            }
          }
        />
        <div
          className="ExecutionPlanBox__DisclosureTriangle-sc-1bacjj9-1 jOyxhL"
          onClick={[Function]}
        />
      </div>
      <div
        className="ExecutionPlanBox__ExecutionPlanBoxName-sc-1bacjj9-3 hFUwrq"
        title="many_materializations_and_passing_expectations"
      >
        many_materializations_and_passing_expectations
      </div>
      <div
        className="ExecutionPlanBox__ExecutionTimeContainer-sc-1bacjj9-2 dabeHx"
        style={
          Object {
            "minWidth": 47.6,
          }
        }
      >
        21 msec
      </div>
    </div>
  </div>
</div>
`;

exports[`renders given materializations and passing expectations 2`] = `
<div
  className="ExecutionPlanBox__ExecutionPlanRowContainer-sc-1bacjj9-5 kaLnAx"
>
  <div
    className="ExecutionPlanBox__ExecutionPlanBoxContainer-sc-1bacjj9-6 beJStN succeeded"
    onClick={[Function]}
    style={
      Object {
        "transitionDelay": "200ms",
      }
    }
  >
    <div
      className="ExecutionPlanBox__ExecutionFinishedFlash-sc-1bacjj9-4 bRVdHa"
      style={
        Object {
          "transitionDelay": "200ms",
        }
      }
    />
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "inline-flex",
        }
      }
    >
      <div
        className="ExecutionPlanBox__ExecutionStateWrap-sc-1bacjj9-0 fipiZp"
        onClick={[Function]}
      >
        <div
          className="ExecutionStateDot-bprv14-0 gPIOyE"
          style={
            Object {
              "transitionDelay": "200ms",
            }
          }
          title="Succeeded"
        />
        <div
          style={
            Object {
              "width": 4,
            }
          }
        />
        <div
          className="ExecutionPlanBox__DisclosureTriangle-sc-1bacjj9-1 jOyxhL"
          onClick={[Function]}
        />
      </div>
      <div
        className="ExecutionPlanBox__ExecutionPlanBoxName-sc-1bacjj9-3 hFUwrq"
        title="many_materializations_and_passing_expectations"
      >
        many_materializations_and_passing_expectations
      </div>
      <div
        className="ExecutionPlanBox__ExecutionTimeContainer-sc-1bacjj9-2 dabeHx"
        style={
          Object {
            "minWidth": 47.6,
          }
        }
      >
        21 msec
      </div>
    </div>
  </div>
</div>
`;
